packages/next/src/build/utils.ts

```diff
@@ -1163,9 +1163,17 @@ export const collectGenerateParams = async (
   const page: string | undefined = segment[0]
   const isClientComponent = isClientReference(mod)
   const isDynamicSegment = /^\[.+\]$/.test(page || '')
-  const { generateStaticParams, getStaticPaths } = mod || {}
+  let { generateStaticParams } = mod || {}
+  const { getStaticPaths } = mod || {}
+
+  // Look for generateStaticParams in template or page if not already found
```

Comment:
Currently it prioritizes exports in layout, then template and lastly page. Except for if the route does not contain a layout, then it prioritizes page.

A better solution would be to check if generateStaticParams is exported from multiple files in the same route, and then throw an error. However this requires all these files in the route to be imported (layout, template, page) making for slower build times. Is it worth the trade-off to get good error messages?